File: MATLAB/optim/estCFO_tb.m
close all
N = 128;
forget_factor =1/2;
ac_cfo_in = 0;

ac_keep2 = [];
% load r_to_cfo_test
load r_tb_for_cfo_est
iter_nums = size(r_tb_for_cfo_est,1);
tb_for_cfo_est = r_tb_for_cfo_est.';
tb_for_cfo_est = tb_for_cfo_est / 2^13;
tb_for_cfo_est = reshape(tb_for_cfo_est, [], 1);
writematrix([real(tb_for_cfo_est), imag(tb_for_cfo_est)], 'data4cfoEst_in.txt', 'Delimiter', 'tab');
ac_cfo_keep = zeros(iter_nums,1);
cfo2piTs = zeros(iter_nums,1);
for iter = 1:iter_nums
    adc_data = r_tb_for_cfo_est(iter,:);

    if iter == 1 
        fist_time_or_not = 1;
    else
        fist_time_or_not = 0;
    end

    ac_cfo = estCFO(adc_data,forget_factor,fist_time_or_not,ac_cfo_in);
    ac_cfo_in = ac_cfo;
    ac_cfo_keep(iter,1)  = ac_cfo;

    cfo2piTs(iter,1) = angle(ac_cfo)/N;

end 
acCFOKeep_out = ac_cfo_keep / 2^26;
writematrix([real(acCFOKeep_out), imag(acCFOKeep_out)], 'acCFOKeep_out.txt', 'Delimiter', 'tab');   

File: HLS/resource_opt1/estCFO_tb.cpp
#include <iostream>
#include <fstream>
#include <complex>
#include "estCFO.hpp"

int main() {
    // Input and output streams
    hls::stream<complex_data_t> adc_data_stream;
    hls::stream<forget_t> forget_factor_stream;
    hls::stream<bool> first_time_stream;
    hls::stream<complex_accum_t> ac_cfo_in_stream;
    hls::stream<complex_accum_t> ac_cfo_out_stream;

    // Load test data from file (replace with your actual data loading)
    std::ifstream dataFile("data4cfoEst_in.txt"); // Assuming data is in a text file
    if (!dataFile.is_open()) {
        std::cerr << "Error opening data file!" << std::endl;
        return 1;
    }
    // Load golden reference data from file
    std::ifstream refFile("acCFOKeep_out.txt"); // Assuming data is in a text file
    if (!refFile.is_open()) {
        std::cerr << "Error opening reference file!" << std::endl;
        return 1;
    }

    // Test parameters
    forget_t forget_factor = 0.5; // Example value
    complex_accum_t ac_cfo_in = {0.0, 0.0}; // Initial CFO estimate
    // Result data array
    complex_float_t ac_cfo_out_ref[ITER_NUM];
    complex_float_t ac_cfo_out_buf[ITER_NUM];

    // Loop through test iterations
    for (int iter = 0; iter < ITER_NUM; ++iter) { //iter<=iter_nums
        bool first_time = (iter == 0);

        // Read ADC data and push to stream
        for (int n = 0; n < 512; ++n) {
            float real_part, imag_part;
            dataFile >> real_part >> imag_part;
            complex_data_t adc_data(real_part, imag_part);
            adc_data_stream.write(adc_data);
        }

        // Push test parameters to streams
        forget_factor_stream.write(forget_factor);
        first_time_stream.write(first_time);
        ac_cfo_in_stream.write(ac_cfo_in);

        // Call the HLS function
        estCFO(adc_data_stream, forget_factor_stream, first_time_stream, ac_cfo_in_stream, ac_cfo_out_stream);

        // Read the output from the stream
        complex_data_t ac_cfo_out = ac_cfo_out_stream.read();
        ac_cfo_out_buf[iter] = {(float)ac_cfo_out.real(), (float)ac_cfo_out.imag()};
        float real_part_ref, imag_part_ref;
        refFile >> real_part_ref >> imag_part_ref;
        ac_cfo_out_ref[iter] = {real_part_ref, imag_part_ref};

        // Print or store the output for verification
        std::cout << "Iteration " << iter << ": ac_cfo_out = " << ac_cfo_out << std::endl;

        // Update ac_cfo_in for the next iteration
        ac_cfo_in = ac_cfo_out;
    }

    // Compare the output with the golden reference
    int err_cnt = 0;
    for (int i = 0; i < ITER_NUM; ++i) {
        if ((std::abs(ac_cfo_out_buf[i].real() - ac_cfo_out_ref[i].real()) > 1e-4) || (std::abs(ac_cfo_out_buf[i].imag() - ac_cfo_out_ref[i].imag()) > 1e-4)) {
            std::cerr << "Mismatch at iteration " << i << ": ac_cfo_out = " << ac_cfo_out_buf[i] << ", expected " << ac_cfo_out_ref[i] << std::endl;
            ++err_cnt;
        }
    }
    if (err_cnt == 0) {
        std::cout << "Test passed!" << std::endl;
    } else {
        std::cout << "Test failed with " << err_cnt << " mismatches!" << std::endl;
    }

    // Close files
    dataFile.close();
    refFile.close();
    return 0;
}
